coriolis/licensing/client.py

```diff
@@ -0,0 +1,188 @@
+# Copyright 2019 Cloudbase Solutions Srl
+# All Rights Reserved.
+
+import json
+import os
+import requests
+
+from coriolis import exception
+from coriolis import utils
+from oslo_log import log as logging
+
+
+LOG = logging.getLogger(__name__)
+
+DEFAULT_LICENSING_PROTOCOL = "http"
+DEFAULT_LICENSING_PORT = 37667
+DEFAULT_LICENSING_API_VERSION = "v1"
+
+RESERVATION_TYPE_REPLICA = "replica"
+RESERVATION_TYPE_MIGRATION = "migration"
+
+
+class LicensingClient(object):
+    """ Class for accessing the Coriolis licensing server API. """
+
+    def __init__(self, base_url, allow_untrusted=False):
+        """ :param base_url: URL for the API service, including scheme """
+        self._base_url = base_url.rstrip('/')
+        self._verify = not allow_untrusted
+
+    @classmethod
+    def from_env(cls):
+        """ Retuns a `LicensingClient` object instatiated using the
+        following env vars:
+        LICENSING_SERVER_BASE_URL="https://10.7.2.3:37667/v1"
+        LICENSING_SERVER_ALLOW_UNTRUSTED="<set to anything>"
+        Returns None if 'LICENSING_SERVER_BASE_URL' is not defined.
+        """
+        base_url = os.environ.get("LICENSING_SERVER_BASE_URL")
+        if not base_url:
+            LOG.warn(
+                "No 'LICENSING_SERVER_BASE_URL' env var present. Cannot "
+                "instantiate licensing client.")
+            return None
+        allow_untrusted = False
+        if os.environ.get("LICENSING_SERVER_ALLOW_UNTRUSTED"):
+            allow_untrusted = True
+
+        # try out client:
+        client = cls(base_url, allow_untrusted=allow_untrusted)
+        try:
```

Comment:
Not much use if you are only going to log if it fails. If something is wrong with the licensing server, it will fail again on next call, which will yield relevant info to the user. Log messages will simply stream by unnoticed until something does wrong.

This would be useful only if we actually send a notification to a monitoring system or an operator, that something is wrong.

Reply:
True, but I was trying to avoid having the conductor crash as much as possible, which is why both this check, and the reservation clearing call are wrapped in a try-except.
Admittedly this call is a bit redundant, so I'll just remove it completely.

Reply:
This call does not prevent the conductor from throwing errors. It just logs if something goes wrong, and nobody will ever notice that log :).